[Compose 페이먼츠 3단계] 이든 미션 제출합니다.

devfeijoa opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 페토 ! step 3 도 잘 부탁드립니다
DM으로 양해를 구했듯이 이번주 테코톡이라 이 상태로 보내는 점 양해 부탁드립니다 🙏 
테코톡 이후에 빠르게 수정하여 다시 리뷰요청 드리겠습니다. 감사합니다 !

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

1. domain 모델에 enum class 인 CardCompany 를 만들었습니다.
그런데 NOT_SELECTED 라는 카드 회사가 선택되지 않은 상태가 과연 domain 인가 UI 인가 한참 고민을 하다가,, 선택되지 않았다는건 domain이 아니라 UI 이라는 결론을 내리고 넣지 않았습니다 !
NOT_SELECTED 같은 상태도 UI에서만 관리하면 되도록CardCompanyUiModel? nullable 로 있도록 했습니다.
이 점에 대해서 페토의 생각이 궁금합니다 🤔

2. 따라서 CardCompanyUiModel에서 해당 카드 회사 이름, 아이콘, 색상을 넣어서 ui 와 domain간의 결합을 낮추도록 노력해봤는데 맞는진 모르겠네요.. 😓 

3. 카드 등록 진입 시점에도 바텀시트가 떠야하고, 카드 모양을 눌렀을 때도 바텀시트가 뜨도록해서 카드 변경을 가능하게 했습니다..! 하지만 요구사항에는 딱히 드러나 있지는 않았어서 이에 대한 페토의 생각이 궁금합니다!

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

devfeijoa committed:
docs: step3 기능 명세서 업데이트

devfeijoa committed:
feat: 카드사 아이콘 추가

devfeijoa committed:
feat: CardCompany enum class 구현

devfeijoa committed:
feat: 카드 회사 UI 모델 및 문자열 리소스 추가

devfeijoa committed:
feat: CardCompanyUiModel과 CardCompany 간의 변환 함수 추가

devfeijoa committed:
feat: CardCompanySelectRow 구현

devfeijoa committed:
feat: 카드 회사 선택 바텀 시트 구현

devfeijoa committed:
feat: 카드 회사 선택 기능 추가

- `NewCardScreen`에 카드 회사 선택 BottomSheet를 추가하고, 선택된 카드 회사를 `PaymentCard`에 표시하도록 수정했습니다.
- `PaymentCard`에 카드 회사 정보를 표시하고, 클릭 시 BottomSheet를 표시하도록 `onCompanyClick` 콜백을 추가했습니다.
- `CardCompanyUiModel`을 `rememberSaveable`로 상태 관리하도록 수정했습니다.

devfeijoa committed:
feat: 카드 회사 선택 바텀 시트 UI 구현

- 카드 회사 선택을 위한 `CardCompanySelectBottomSheet` 컴포저블을 추가했습니다.
- 사용자가 카드 회사를 선택하거나 바텀 시트를 닫을 수 있는 기능을 제공합니다.
- `CardCompanyUiModel`을 사용하여 각 카드 회사의 정보를 표시합니다.
- Preview를 추가하여 UI를 쉽게 확인할 수 있도록 했습니다.

devfeijoa committed:
feat: 카드사에 따라 다른 카드 이름 표시 기능 추가

- `PaymentCard` 컴포저블에 카드 회사 정보를 표시합니다.
- 카드 회사 정보가 있으면 해당 회사 이름을 표시하고, 없으면 기본 회색 배경을 표시합니다.
- 카드 회사를 클릭하면 `onCompanyClick` 콜백을 호출합니다.

devfeijoa committed:
feat: 카드사에 따라 다른 카드 이름 및 배경색 표시 기능 추가

devfeijoa committed:
refactor: 카드 목록 화면에서 카드 추가 버튼 표시 조건 수정

devfeijoa committed:
refactor: RegisteredCard 컴포넌트 UI 수정

- 카드사 이름이 없는 경우를 고려하여 UI를 수정했습니다.
- 카드 정보 텍스트의 `lineHeight`를 1.em으로 설정했습니다.
- 카드사 이름, 카드 번호, 소유자 이름, 만료일의 위치 및 정렬을 조정했습니다.

devfeijoa commented on app/src/main/java/woowacourse/payments/ui/component/CardCompanySelectBottomSheet.kt:
```diff
@@ -0,0 +1,61 @@
+    val sheetState = rememberModalBottomSheetState()
+    var selectedCompany by remember { mutableStateOf<CardCompanyUiModel?>(null) }
+
+    LaunchedEffect(selectedCompany) {
+        selectedCompany?.let {
+            sheetState.hide()
+            onDismissRequest()
+        }
+    }
+
```
이렇게 구현해보았는데, hide를 하고 `onDismissRequest()`를 하는게 정확하게 바텀 시트를 닫는건진 확실하지 않습니다.. 😢 
굳이 hide()를 안 해도 되었을까요? 이에 대한 페토의 생각이 궁금합니다 🤔

chanho0908 replied:
`onDismissRequest()`를 하면 `NewCardScreen`의 `showBottomSheet` 상태를 변경시키고 있군요 !
이든께서 작성해주신 코드를 보면 `showBottomSheet` 상태에 따라 `CardCompanySelectBottomSheet`를 아에 렌더링 하지 않고 있어요 ! 
```
if (showBottomSheet) {
  CardCompanySelectBottomSheet(
      onCompanyClick = { company ->
          selectedCardCompany = company
          showBottomSheet = false
      },
      onDismissRequest = {
          showBottomSheet = false
      },
  )
}
```

저는 이 부분에서 실수를 하나 했는데요 😢 렌더링 여부는 변경하지 않고 `hide()` 만 하니까 바텀시트가 내려가도 화면이 터치가 되지 않더라구요 😅

### ❌ 틀릴 수 있음 주의 !! ❌  
제가 나름대로 문제를 분석해본 결과 `hide()`는 단순히 가시성만 변경하기 때문에 카드사를 선택하기 전까지 바텀시트 외부 영역을 터치할 수 없게 했던 옵션이 유지됐던 것 같아요

때문에 단순히 가시성을 변경하는 것 뿐만 아니라 이든이 구현해주신 것 처럼 바텀시트 자체를 렌더링 되지 않게 했더니 문제가 해결됐습니다 👍 

이 부분은 이든이 저보다 잘 만들어주셨네요 ㅎㅎ 👍 👍 💯 
</hr>

> onDismissRequest()를 하는게 정확하게 바텀 시트를 닫는건진 확실하지 않습니다

바텀시트를 렌더링 하지 않기 때문에 닫는다...? 라고 표현하긴 애매하지만 암튼 사라지는건 맞음 😆

> 굳이 hide()를 안 해도 되었을까요?

뇌피셜이지만 렌더링을 아에 안하니까 안해도 될듯 ...?
</hr>

트러블 슈팅하며 기록한 [문서](https://chanho0908.notion.site/Compose-ModalBottomSheet-26b0f436649f8004ab03f0d18fd8b42f)를 참고차 남길게요!

chanho0908 commented on app/src/main/java/woowacourse/payments/domain/CardCompany.kt:
```diff
@@ -0,0 +1,12 @@
+package woowacourse.payments.domain
+
+enum class CardCompany {
+    BC,
+    SHINHAN,
+    KAKAO,
+    HYUNDAI,
+    WOORI,
+    LOTTE,
+    HANA,
+    KB,
+}
```
> domain 모델에 enum class 인 CardCompany 를 만들었습니다.
그런데 NOT_SELECTED 라는 카드 회사가 선택되지 않은 상태가 과연 domain 인가 UI 인가 한참 고민을 하다가,, 선택되지 않았다는건 domain이 아니라 UI 이라는 결론을 내리고 넣지 않았습니다 !
NOT_SELECTED 같은 상태도 UI에서만 관리하면 되도록 CardCompanyUiModel? nullable 로 있도록 했습니다.
이 점에 대해서 페토의 생각이 궁금합니다 🤔

좋은 고민이 있었군요 ! 저도 이든과 같은 고민이 있었어요

우린 카드 등록 프로그램 만들기라는 미션을 하고있고 이번 요구사항에선 "카드사"라는 새로운 요구사항이 추가되었죠
그리고 우리는 등록 가능한 카드사들을 정의해 우리만의 약속으로 만들었어요. 즉, 도메인을 만든거죠 !

저 또한 이러한 규칙속에서 "카드사가 선택되지 않음"이라는 상태는 존재할 수 없다고 생각했슴니다
현실 세계에서만 생각해도, 카드사가 선택되지 않았는데 카드를 등록하는건 애초에 말이 안 되니까요 😆 

NOT_SELECTED 상태를 UI에서만 관리하는 것도 너무 공감되는 말이에요 저도 UI에서 관리해야한다고 생각해요 👍 
다만 NOT_SELECTED를 표현함에 있어서 null을 선택하신건 조금 아쉬운 것 같아요

물론 null이 무조건 나쁘다는건 아닙니다 ! 저도 컴포즈 미션을 하기 전까지 무언가 없다는 상태를 null 로 표현하곤 했어요
하지만 컴포즈 미션을 하면서 null이 과연 우리가 의도한 상태를 표현하기에 최선의 수단인가? 를 고민해봤을 때 다소 어렵다는 생각을 했어요

이유는 다른 개발자가 `null Type`을 보고 이 `null`이 의미하는 것이 정확히 무엇인지 의도 파악을 쉽게 할 수 있을까? 에 대한 고민을 해봤는데요, 결론적으로 사람마다 다르게 해석될 여지가 매우 많다고 생각해요

`var selectedCompany by remember { mutableStateOf<CardCompanyUiModel?>(null) }`
+ 카드사가 선택되지 않은 상탠가?
+ 카드사는 선택하지 않아도 되는 것인가 ? (심지어 카드사 선택 안해도 등록할 수 있음 ☠️)
+ 이 회사는 카드사를 등록하지 않아도 카드를 등록할 수 있네 ?

등등 다양하게 해석될 여지가 너무 많은 것 같아요 그래서 이번에 저는 상태를 표현하는 방법에 대해 배우게 되었는데 이든께서 한번 고민해보시고 적용해보면 좋을 것 같아요

chanho0908 replied:
### 1. 카드사 선택에 따른 UI 상태 만들기
<details>
  <summary>코드</summary>

```kotlin
sealed interface CardCompanyUiState {
    data object NotSelected: CardCompanyUiState
    data class Selected(val company: CardCompanyUiModel): CardCompanyUiState
}
```
</details>

### 2. 카드사 선택 바텀시트 컴포저블
<details>
  <summary>코드</summary>

```kotlin
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CardCompanySelectBottomSheet(
    modifier: Modifier = Modifier,
    onCompanyClick: (CardCompanyUiModel) -> Unit,
    onDismissRequest: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState()
    var cardCompanyState by remember { mutableStateOf<CardCompanyUiState>(CardCompanyUiState.NotSelected) }
    
    LaunchedEffect(cardCompanyState) {
        when (cardCompanyState) {
            CardCompanyUiState.NotSelected -> { 
                // 바텀시트 열기
            }
            is CardCompanyUiState.Selected -> { 
                // 바텀시트 닫기
            }
        }
    }
    
    ModalBottomSheet(
        modifier = modifier,
        sheetState = sheetState,
        onDismissRequest = onDismissRequest,
        containerColor = Color.White,
    ) {
        CardCompanySelectRow(
            modifier = Modifier.navigationBarsPadding(),
            onCompanyClick = { company ->
                cardCompanyState = CardCompanyUiState.Selected(company)
                onCompanyClick(company)
            },
        )
    }
}
```
</details>

chanho0908 replied:
이렇게만 하면 될 줄 알았는데 계속 화면이 회전 됐을 때 상태가 유지되지 않아 코드를 좀 더 살펴보니 상태를 관리하는 곳이 여러군데였군요 !

```
@Composable
fun NewCardScreen(
    onBackClick: () -> Unit = {},
    onSaveClick: (CardUiModel) -> Unit = {},
) {
    var cardNumber by remember { mutableStateOf("") }
    var cardExpiryDate by remember { mutableStateOf("") }
    var cardHolderName by remember { mutableStateOf("") }
    var cardPassword by remember { mutableStateOf("") }

    var selectedCardCompany by rememberSaveable { mutableStateOf<CardCompanyUiModel?>(null) }
    var showBottomSheet by remember { mutableStateOf(false) }
    val context = LocalContext.current

    LaunchedEffect(Unit) {
        showBottomSheet = true
    }

```

이 부분이 원인이었던 것 같아요 🤔 이든께서 컴포즈의 상태 관리와 친해질 수 있도록 처음부터 이야기해볼게요

저희 2단계 강의 자료를 보면 Stateful vs Stateless에 관한 내용이 있어요 !

<img width="935" height="459" alt="image" src="https://github.com/user-attachments/assets/3e1111f7-d51e-4ce9-99be-97cee84625e4" />
강의 자료에 나와있는 것 처럼 `Stateful` 이란, `remember`나 `mutableStateOf` 상태를 가진 컴포저블을 의미해요
이러한 컴포저블들의 단점은 테스트가 어렵다는 점인데요, 이 말은 Preview를 통해 컴포저블의 다양한 상태를 미리 확인할 수 없다는 것 또한 의미해요  ☹️

반대로 `Stateless`는 이러한 상태가 없는 컴포저블을 의미하구요 !

저는 컴포즈 미션을 진행하면서 상태를 외부에서 주입받아야 프리뷰로 확인하기 편하길래 그냥 다 외부에서 주입받도록 만들었는데 이게 Stateful vs Stateless이란 개념이더라구요 저도 몰랐음 !

저는 이런 `Stateful`한 컴포저블의 장점을 테스트코드를 만들면서 정말 크게 느꼈기 때문에 이든에게 테스트 테스트 노래를 불렀던거에요 🤣 테스트까진 어렵더라도, 프리뷰를 통해서 이든이 `Stateless`한 컴포저블의 장점을 느껴보셨으면 좋겠어요 👍

chanho0908 replied:
그렇다면 이러한 Stateless한 컴포저블은 상태를 어디서 관리하는가? 의문이 들것 같아요 🤔 

그래서 이 때 사용하는 개념이 상태 호이스팅입니다.
대공백의 기가막힌 설명을 차용하자면, **상태를 호잇! 하고 위로 올리는 거에요** 

그럼 상태를 어디까지 올려야 하나? 상태를 전부다 호이스팅 해야하나? 같은 고민들을 처음부터 하게되면 너무 어려울 것 같아 저는 일단은 어쩌고 저쩌고 Screen 까지 올린다고 생각하고 있습니다. (사실 저도 아직은 잘 모르겠음 )

<img width="954" height="649" alt="image" src="https://github.com/user-attachments/assets/e56211e1-31e3-4c01-a24a-0bba1514259c" />
<img width="935" height="380" alt="image" src="https://github.com/user-attachments/assets/500c921b-4c29-4c3e-ba9f-88d78edaef77" />

이렇게 상태를 호이스팅하면 하위 컴포저블은 외부에서 값을 주입받기 때문에 프리뷰나 테스트가 편해지고 상태를 한곳에서 관리할 수 있기 때문에 상태 관리가 보다 용이해 진답니다 😄 

최종적으로 정리해보면 다음과 같아요

+ Stateful이란 `remember`, `mutableStateOf` 같은 상태를 자기 자신이 가진 컴포저블
+ Stateful하면 테스트나 프리뷰가 어렵다. Stateless 짱짱맨 😎
+ Stateless한 컴포저블을 만들기 위해선 상태 호이스팅을 사용하자
+ 어디까지 ? 지금은 잘 모르겠으니 아묻따 어쩌고 저쩌고 Screen 까지 해보자

리뷰가 여기에만 너무 쌓이는 것 같아서 이후 내용은 `CardCompanySelectBottomSheet`에서 이어갈게요

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/CardCompanySelectBottomSheet.kt:
```diff
@@ -0,0 +1,61 @@
+import androidx.compose.ui.tooling.preview.Preview
+import woowacourse.payments.ui.model.CardCompanyUiModel
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun CardCompanySelectBottomSheet(
+    modifier: Modifier = Modifier,
+    onCompanyClick: (CardCompanyUiModel) -> Unit,
+    onDismissRequest: () -> Unit,
+) {
```
틀린그림 찾기 한번 해보시겠어요 ?

devfeijoa replied:
파라미터 순서 가이드 라인에 따라 수정하였습니다 👍 
8a9780ecad0666cddc2668abcd69ea9f9c33d657

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/CardCompanySelectBottomSheet.kt:
```diff
@@ -0,0 +1,61 @@
+            sheetState.hide()
+            onDismissRequest()
+        }
+    }
```
앞서 제가 요런 코드를 먼저 제안드렸어요 ! 아쉽게도 이 코드는 `Stateful`해서 개선이 필요해보여요 ! 
이 아이를 Stateless하게 만드려면 어떻게 해야할까요 ? 🤔 
<details>
  <summary>코드</summary>

```kotlin
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CardCompanySelectBottomSheet(
    modifier: Modifier = Modifier,
    onCompanyClick: (CardCompanyUiModel) -> Unit,
    onDismissRequest: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState()
    var cardCompanyState by remember { mutableStateOf<CardCompanyUiState>(CardCompanyUiState.NotSelected) }
    
    LaunchedEffect(cardCompanyState) {
        when (cardCompanyState) {
            CardCompanyUiState.NotSelected -> { 
                // 바텀시트 열기
            }
            is CardCompanyUiState.Selected -> { 
                // 바텀시트 닫기
            }
        }
    }
    
    ModalBottomSheet(
        modifier = modifier,
        sheetState = sheetState,
        onDismissRequest = onDismissRequest,
        containerColor = Color.White,
    ) {
        CardCompanySelectRow(
            modifier = Modifier.navigationBarsPadding(),
            onCompanyClick = { company ->
                cardCompanyState = CardCompanyUiState.Selected(company)
                onCompanyClick(company)
            },
        )
    }
}
```
</details>

이든의 코드와 제가 제안드린 코드를 봤을 때 공통적으로 상태라고 할 수 있는 부분은 요기 같아요
```
val sheetState = rememberModalBottomSheetState()
var selectedCompany by remember { mutableStateOf<CardCompanyUiModel?>(null) }
```

이 상태를 외부에서 주입받으면 Stateless하게 만들 수 있을 것 같아요
<details>
  <summary>힌트</summary>

```kotlin
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CardCompanySelectBottomSheet(
    sheetState: SheetState,
    modifier: Modifier = Modifier,
    onCompanyClick: (CardCompanyUiState) -> Unit,
    onDismissRequest: () -> Unit,
) {
    ModalBottomSheet(
        modifier = modifier,
        sheetState = sheetState,
        onDismissRequest = onDismissRequest,
        containerColor = Color.White,
    ) {
        CardCompanySelectRow(
            modifier = Modifier.navigationBarsPadding(),
            onCompanyClick = { company ->
                onCompanyClick(CardCompanyUiState.Selected(company))
            },
        )
    }
}


```
</details>

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/CardCompanySelectRow.kt:
```diff
@@ -0,0 +1,86 @@
+@Composable
+fun CardCompanySelectRow(
+    modifier: Modifier = Modifier,
+    onCompanyClick: (CardCompanyUiModel) -> Unit,
```
틀린그림 찾기 ~~

devfeijoa replied:
파라미터 순서 가이드 라인에 따라 수정하였습니다 👍 
8a9780ecad0666cddc2668abcd69ea9f9c33d657

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/CardCompanySelectRow.kt:
```diff
@@ -0,0 +1,86 @@
+    }
+}
+
+@Composable
+private fun CardCompanyItem(
+    modifier: Modifier = Modifier,
+    company: CardCompanyUiModel,
+    onClick: () -> Unit,
+) {
```
틀린 그림 찾기 😤

devfeijoa replied:
파라미터 순서 가이드 라인에 따라 수정하였습니다 👍 
8a9780ecad0666cddc2668abcd69ea9f9c33d657

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -35,8 +39,14 @@ fun NewCardScreen(
     var cardPassword by remember { mutableStateOf("") }
 
+    var selectedCardCompany by rememberSaveable { mutableStateOf<CardCompanyUiModel?>(null) }
+    var showBottomSheet by remember { mutableStateOf(false) }
```
오!! 이렇게 하면 카드사를 다시 선택할 수 있군요 👍 💯  저는 생각 못한 부분인데 든든한태 하나 배워갑니다 ☺️

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -35,8 +39,14 @@ fun NewCardScreen(
     var cardHolderName by remember { mutableStateOf("") }
     var cardPassword by remember { mutableStateOf("") }
 
+    var selectedCardCompany by rememberSaveable { mutableStateOf<CardCompanyUiModel?>(null) }
```
앞서 제안드린 상태 호이스팅을 한다면 이렇게 호이스팅해서 사용할 수 있을 것 같아요 !

```
var cardCompanyState by rememberSaveable {
        mutableStateOf<CardCompanyUiState>(
            CardCompanyUiState.NotSelected
        )
    }
val sheetState = rememberModalBottomSheetState()
var showBottomSheet by rememberSaveable {
  mutableStateOf(cardCompanyState is CardCompanyUiState.NotSelected)
}
```

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/model/CardCompanyUiModel.kt:
```diff
@@ -0,0 +1,58 @@
+    ),
+}
+
+fun CardCompany.toUiModel(): CardCompanyUiModel? = CardCompanyUiModel.entries.find { it.name == this.name }
+
+fun CardCompanyUiModel.toDomain(): CardCompany? = CardCompany.entries.find { it.name == this.name }
```
이 아이들도 nullable 한게 조금 아쉬운 것 같아요 🤔 

논리적으로 예외가 절대 발생 할일이 없다면 그냥 에러가 발생하게 만들어서 Non Null 타입으로 만드는건 어떻게 생각할까요 ?

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/model/CardCompanyUiModel.kt:
```diff
@@ -0,0 +1,58 @@
+    BC(
+        companyName = R.string.card_company_bc,
+        companyIcon = R.drawable.ic_bc,
+        color = Color(0xFFF04651),
```
Compose의 Color.kt 파일을 사용하면 색상을미리 정의해놓고 사용할 수 도 있을 것 같아요 👍

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,32 +1,65 @@
-fun PaymentCard(modifier: Modifier = Modifier) {
+fun PaymentCard(
+    modifier: Modifier = Modifier,
+    cardCompany: CardCompanyUiModel? = null,
```
앞서 제안드린 `CardCompanyUiState`를 받으면 코드가 더 개선될 것 같은데 이든께선 어떻게 생각하실까요 ? 🤔

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,32 +1,65 @@
+    cardCompany: CardCompanyUiModel? = null,
+    onCompanyClick: (() -> Unit)? = null,
+) {
+    val cardColor = cardCompany?.color ?: Grey100
```
<details>
  <summary>힌트</summary>

```kotlin
val cardColor = when (cardCompanyUiState) {
        CardCompanyUiState.NotSelected -> Grey100
        is CardCompanyUiState.Selected -> cardCompanyUiState.company.color
    }
```
</details>

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,32 +1,65 @@
             modifier
                 .shadow(8.dp)
                 .size(width = 208.dp, height = 124.dp)
+                .then(clickableModifier)
```
`then`??!! 오 이건 무엇인가요 ? 저는 처음보는건데 역시 Modifier 테코톡 권위자 든든 ... 👍 👍 💯

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,32 +1,65 @@
                     shape = RoundedCornerShape(5.dp),
                 ),
     ) {
+        cardCompany?.let { company ->
+            Text(
+                text = stringResource(id = company.companyName),
+                color = Color.White,
+                fontSize = 12.sp,
+                fontWeight = FontWeight.W500,
+                modifier =
+                    Modifier
+                        .align(Alignment.BottomStart)
+                        .padding(start = 14.dp, bottom = 85.dp),
+            )
+        }
```
<details>
  <summary>힌트</summary>

```kotlin
if (cardCompanyUiState is CardCompanyUiState.Selected) {
            Text(
                text = stringResource(id = cardCompanyUiState.company.companyName),
                color = Color.White,
                fontSize = 12.sp,
                fontWeight = FontWeight.W500,
                modifier =
                    Modifier
                        .align(Alignment.BottomStart)
                        .padding(start = 14.dp, bottom = 85.dp),
            )
        }
```
</details>

chanho0908 reviewed (changes requested):
안녕하세요 든든! 테코톡 기간까지 겹치면서 미션을 수행해내기 힘드셨을탠데 너무 고생 많으셨습니다 🥺

우테코를 하면서 이든과 함께 미션을 해처 나갔던 날이 정말 많았는데 오랜만에 제 도움 없이 이든 스스로 해내신거 정말 멋지다고 생각합니다 👍  

최고에요 든든 😁 눈물 아주 살짝 찔끔 흘려도 괜찮죠 ? 🥲

리뷰를 어떤식으로 남겨야 이든에게 도움이 될지 정말 많이 고민했어요 🤔 
사실 아직도 잘 모르겠어서 그냥 마음 가는대로 해봤는데요, 

이든이 미션을 기깔나게 잘 구현해주셨으니 수업에서 배운 내용중에 아직 적용되지 않은 내용 위주로 의견을 남겼어요 ! 

리뷰를 고봉밥으로 남긴만큼 가이드를 조금 제안드리고 싶은데요, 제가 남긴 리뷰는 모든 내용을 적용하라고 남긴것이 아닙니다. 

이번 미션 기간이 아니라도 추후 이든께서 학습하시는 과정에서 도움이 되고자 하여 최대한 많은 내용을 남겼으니 이든께서 적용 가능한 부분까지만 적용해주세요

이든께서 스스로 고민해보실 수 있도록 어떤 곳엔 모든 코드를, 어떤 곳엔 힌트를 남겼고 아에 의견을 드리지 않은 부분이 있습니다. 힌트라고 남긴곳은 이든께서 먼저 고민해보시고 모르겠다 싶으면 보시는걸 추천드려요 ! 

그리고 저 또한 아직 컴포즈가 미숙한지라 이든께서 더 좋은 방법을 찾으신다면 편하게 이든의 의견대로 구현해주셔도 괜찮습니다 🙂 

함께 성장해나가는 동료인 만큼, 이든과 함께 더 나은 답을 찾아나가고 싶습니다 ☺️

언제나 그렇듯, 도움이 필요하거나 함께 고민해보고 싶은 내용이 있다면 편하게 DM 주세요 ! 

고생 많으셨습니다 든든  🫡

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/model/CardCompanyUiModel.kt:
```diff
@@ -0,0 +1,58 @@
+import woowacourse.payments.R
+import woowacourse.payments.domain.CardCompany
+
+enum class CardCompanyUiModel(
+    @StringRes val companyName: Int,
+    @DrawableRes val companyIcon: Int,
+    val color: Color,
+) {
```
> 따라서 CardCompanyUiModel에서 해당 카드 회사 이름, 아이콘, 색상을 넣어서 ui 와 domain간의 결합을 낮추도록 노력해봤는데 맞는진 모르겠네요.. 😓

이든께서 도메인에 필요한 정보와 ui에 필요한 정보를 구분하는데 있어서 많은 고민이 있으셨을 것 같아요 🤔 
말씀해주신 것 처럼 지금의 방식이 맞다 틀리다를 고민해본다면, 저는 맞다고 생각해요

다른 사람이 보기에 오답이라고 생각할 수 있는 코드도 이든께서 깊이 고민하시며, 이든께서 하실 수 있는 최선의 결정을 내리셨다면 그게 정답이라고 생각합니다 👍 

또한 리뷰어로써 제 의견을 조금 남기자면, 놀랍게도 사실 저도 미션 초기엔 이 코드와 매우 비슷하게 만들었어요
이렇게 하면 `CardCompanyUiModel.entries.forEach `만 하면 내부 프로퍼티를 통해 일관적인 UI 만들 수 있으니까요 !

다만 저는 해당 코드를지우게 되었는데, 저는 Domain에 있는 `CardCompany`가 하드 코딩된 카드사명을 가지고 있었어요
때문에 만약 서버에서 카드사명을 내려준다면, 도메인에 하드 코딩된 카드사명은 필요가 없어지기 때문에 Domain이
data Layer의 변경으로 인해 수정되는, 외부 환경으로부터 순수하지 않은 Domain이라고 생각해서 제거했습니다.

다시 생각해보니 저는 `CardCompany`를 지금 이든이 만드신 것 처럼 도메인의 `CardCompany`와 Ui와 분리할 생각은 못했네요!! 😱 이든 덕에 새로운 관점을 배워갑니다 👍 👍 👍  ☺️

devfeijoa committed:
refactor: 함수 파라미터 순서 가이드 라인에 맞게 변경

devfeijoa committed:
refactor: 불필요한 import 제거

devfeijoa committed:
refactor: 카드사 선택 BottomSheet 및 카드 추가 화면 개선

- `CardCompanySelectBottomSheet`에서 `LaunchedEffect` 및 관련 상태 로직을 제거했습니다.
- `NewCardScreen`에서 `LocalContext` 사용을 제거하고, `CardCompanySelectBottomSheet` 호출 시 불필요한 파라미터를 삭제했습니다.

devfeijoa committed:
refactor: `CardCompanySelectBottomSheet`에서 `selectedCompany` 상태 제거

- `CardCompanySelectBottomSheet` 컴포저블에서 `selectedCompany` 관련 로직을 제거했습니다.
- 카드사 선택 시 `onCompanyClick` 콜백을 통해 선택된 카드사를 `NewCardScreen`으로 전달하도록 수정했습니다.
- `NewCardScreen`에서 `selectedCardCompany` 상태를 관리하고, 바텀 시트의 표시 여부를 제어합니다.
